### PR TITLE
[hail] Add ArrayZip IR node.

### PR DESCRIPTION
**Pull request by tpoterba**
Needs a bit more infrastructure before this can help deforest several
more lowered IRs.

timings:
```
$ hail-bench compare /tmp/pre.json /tmp/post.json
                                              Name      Ratio    Time 1    Time 2
                                              ----      -----    ------    ------
                     matrix_table_array_arithmetic      81.2%     9.967     8.095
matrix_table_nested_annotate_rows_annotate_entries      23.6%    31.287     7.369
```

### CODE REVIEW

**Comment by akotlar** [comment]
So cool

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 1]
```diff
@@ -141,6 +141,8 @@ object Simplify {
 
     case x@ArrayMap(NA(_), _, _) => NA(x.typ)
 
+    case ArrayZip(as, names, body, _) if as.length == 1 => ArrayMap(as.head, names.head, body)
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: You might also do
```scala
case ArrayMap(ArrayZip(as, names, body1, b), name, body2) =>
  ArrayZip(as, names, Let(name, body1, body2), b)
```
to allow optimizations across `body1` and `body2`.

If one of the `as` is an `ArrayZip`, we can fuse them, but I'm not sure how to write that.

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala [hunk 2]
```diff
@@ -251,6 +251,9 @@ object TypeCheck {
       case x@ArrayMap(a, name, body) =>
         assert(a.typ.isInstanceOf[TStreamable])
         assert(x.elementTyp == body.typ)
+      case x@ArrayZip(as, names, body, _) =>
+        assert(as.length == names.length)
+        assert(as.forall(_.typ.isInstanceOf[TStreamable]))
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala]: Should this check `x.elementTyp == body.typ` like `ArrayMap` does? Or should `TypeCheck` not overlap with `InferType`?

File: hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala [hunk 3]
```diff
@@ -497,16 +497,24 @@ object LowerMatrixIR {
...
+          loweredChild,
+          InsertFields(
+            Ref("row", rt),
+            FastSeq((entriesFieldName, ArrayZip(
+              FastIndexedSeq(
+                GetField(Ref("row", rt), entriesFieldName),
+                GetField(Ref("global", gt), colsFieldName)),
+              FastIndexedSeq("g", "sa"),
+              Subst(lower(newEntries, ab), BindingEnv(Env(
+                "global" -> SelectFields(Ref("global", gt), child.typ.globalType.fieldNames),
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/LowerMatrixIR.scala]: We used to do this, but it caused problems, possibly due to some weird interaction between shadowing and optimization. I'd have to dig in to remember exactly why.

File: hail/src/main/scala/is/hail/expr/ir/EmitStream.scala [hunk 4]
```diff
@@ -427,6 +567,32 @@ object EmitStream {
...
           }
 
+        case ArrayZip(as, names, body, behavior) =>
+          val streams = as.map(emitStream(_, env))
+          val childEltTypes = as.map(_.pType.asInstanceOf[PStreamable].elementType)
+
+          EmitStream.zip[Any](streams, behavior, { (xs, k) =>
+            val mv = names.zip(childEltTypes).map { case (name, t) =>
+              val ti = typeToTypeInfo(t)
+              val eltm = fb.newField[Boolean](name + "_missing")
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EmitStream.scala]: that should work, yeah.

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: good call. Fusing zips doesn't seem hard but will leave that for another PR.

File: hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala]: sure, good thing to check here I guess. It's a bit unclear what's the right code for typecheck vs infer.

File: hail/src/main/scala/is/hail/expr/ir/EmitStream.scala [hunk 5]
```diff
@@ -427,6 +567,32 @@ object EmitStream {
...
+              val eltv = fb.newField(name)(ti)
+              (t, ti, eltm, eltv)
+            }
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/EmitStream.scala]: got rid of `mv` per your other comment.

File: hail/python/hail/ir/ir.py [hunk 6]
```diff
@@ -998,6 +998,42 @@ def renderable_bindings(self, i, default_value=None):
...
+        self.names = names
+        self.body = body
+        self.behavior = behavior
+
+    @typecheck_method(children=IR)
+    def copy(self, *children):
+        return ArrayZip(children[:-1], self.names, children[-1], self.behavior)
+
+    def head_str(self):
+        return escape_id(self.behavior) + ' ({})'.format(' '.join(map(escape_id, self.names)))
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/python/hail/ir/ir.py]: I'm a big fan of f-strings, just not sure what to think about having 2 different styles in the same file. I'll change it.

[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/ir/ir.py]: I think this is the kind of thing we can say "This is the style we want to move to. Change any old-style code when you happen to touch it."

**Comment by tpoterba** [comment]
I got a bad local variable compilation error without the fields for the bindings. I suspect this is a locals/fields problem elsewhere but don't want to debug right now, so reverted.

**Pull request merged**